src/NimBLERemoteCharacteristic.cpp

```diff
@@ -63,44 +63,41 @@ NimBLERemoteCharacteristic::~NimBLERemoteCharacteristic() {
  * @brief Callback used by the API when a descriptor is discovered or search complete.
  */
 int NimBLERemoteCharacteristic::descriptorDiscCB(
-    uint16_t conn_handle, const ble_gatt_error* error, uint16_t chr_val_handle, const ble_gatt_dsc* dsc, void* arg) {
+    uint16_t connHandle, const ble_gatt_error* error, uint16_t chrHandle, const ble_gatt_dsc* dsc, void* arg) {
     int rc = error->status;
     NIMBLE_LOGD(LOG_TAG, "Descriptor Discovery >> status: %d handle: %d", rc, (rc == 0) ? dsc->handle : -1);
 
-    auto              filter     = (desc_filter_t*)arg;
-    auto              pTaskData  = (NimBLETaskData*)filter->task_data;
-    const auto        pChr       = (NimBLERemoteCharacteristic*)pTaskData->m_pInstance;
-    const NimBLEUUID* uuidFilter = filter->uuid;
+    auto       filter    = (desc_filter_t*)arg;
+    auto       pTaskData = (NimBLETaskData*)filter->task_data;
+    const auto pChr      = (NimBLERemoteCharacteristic*)pTaskData->m_pInstance;
+    const auto uuid      = filter->uuid; // UUID to filter for
 
-    if (pChr->getHandle() != chr_val_handle) {
-        return 0; // Descriptor not for this characteristic
+    // Error/done, wrong characteristic, or no-match descriptor
+    if (rc != 0 || pChr->getHandle() != chrHandle
+      || (uuid && ble_uuid_cmp(uuid->getBase(), &dsc->uuid.u))) {
+        goto Done;
     }
+    // Results added until rc != 0
+    pChr->m_vDescriptors.push_back(new NimBLERemoteDescriptor(pChr, dsc));
+    // Return BLE_HS_EDONE if the descriptor was found, stop the search
+    rc = !!uuid * BLE_HS_EDONE;
 
-    if (rc == 0) {
-        if (uuidFilter != nullptr) {
-            if (ble_uuid_cmp(uuidFilter->getBase(), &dsc->uuid.u) == 0) {
-                rc = BLE_HS_EDONE; // Found the descriptor, stop the search
-            } else {
-                return 0; // Not the descriptor we are looking for
-            }
-        }
-
-        pChr->m_vDescriptors.push_back(new NimBLERemoteDescriptor(pChr, dsc));
-        return 0;
+Done:
+    if (rc != 0) {
+        NimBLEUtils::taskRelease(*pTaskData, rc);
     }
-
-    NimBLEUtils::taskRelease(*pTaskData, rc);
     NIMBLE_LOGD(LOG_TAG, "<< Descriptor Discovery");
     return rc;
 }
 
 /**
  * @brief Populate the descriptors (if any) for this characteristic.
- * @param [in] the end handle of the characteristic, or the service, whichever comes first.
+ * @param [in] uuidFilter The end handle of the characteristic, or the service, whichever comes first.
+ * @param [out] out Pointer to get the descriptor if found.
+ * @return True if successfully retrieved, success = 0 or BLE_HS_EDONE.
  */
-bool NimBLERemoteCharacteristic::retrieveDescriptors(const NimBLEUUID* uuidFilter) const {
+bool NimBLERemoteCharacteristic::retrieveDescriptors(const NimBLEUUID* uuidFilter, NimBLERemoteDescriptor* out) const {
```

Comment:
Maybe instead of adding a parameter we should just add a `NimBLERemoteDescriptor*` to `desc_filter_t` and pass a pointer to the filter as the parameter to this. Would require creating the instance of the filter in `getDescriptor` instead of course?